### DISCUSSION TITLE
Show and Tell Category?  Ghostty and Fzf love!

### DISCUSSION THREAD

**Discussion by pkazmier** [ghostty-org/ghostty#3527] [Q&A] [answered]
Could we get a GitHub Discussions "Show and Tell" category alongside the Q&A, Ideas, and Announcements? At the moment, I don't know where else to post this for others that might be interested. Thanks!

## Ghostty and Fzf
I wanted to share how I use `fzf` to switch between different Ghostty configuration snippets on my Mac. I like to change fonts multiple times in a day, so being able to do so quickly is important to me. Using `fzf` and a little function I wrote called `gc` (**g**hostty **c**onfig) in my `.zshrc`, I can do things like this:

https://github.com/user-attachments/assets/9fc89323-5d61-4ca5-8f4b-cb89f2666306

## How it works

Here is my main Ghostty config directory. The `font`, `theme`, and `opacity` directories contain the Ghostty configuration snippets that are presented in `fzf`:
```
.
├── config
├── font
│   ├── 0xproto
│   ├── berkeley-mono
│   ├── cascadia-code
│   ├── codelia
│   ├── comic-code
│   ├── commit-mono
│   ├── ellograph
│   ├── ibm-plex
│   ├── intel-one-mono
│   ├── jetbrains-mono
│   ├── m-plus
│   ├── menlo
│   ├── monaspace-argon
│   ├── monaspace-krypton
│   ├── monaspace-neon
│   ├── monaspace-xenon
│   ├── monolisa
│   ├── operator-mono
│   ├── pragmata-pro
│   ├── roboto-mono
│   ├── sf-mono
│   ├── source-code-pro
│   ├── sudo
│   ├── ubuntu-mono
│   └── victor
├── opacity
│   ├── 100%
│   ├── 70%
│   ├── 75%
│   ├── 80%
│   ├── 85%
│   ├── 90%
│   └── 95%
└── theme
    ├── catppuccin-mocha
    ├── default
    ├── gruvbox-material
    ├── tokyonight-moon
    └── tokyonight-night
```

The contents of those configuration files contain small snippets of Ghostty configuration. For example, here are the contents of `font/berkeley-mono`:
```
font-family = TX-02
font-style = Regular
font-style-bold = ExtraBold
font-style-italic = Oblique
font-style-bold-italic = ExtraBold Oblique
font-feature = ss02
adjust-cell-height = 7
adjust-cursor-height = 7
```

And, here are the contents of `font/menlo`:
```
font-family = Menlo
adjust-cell-height = 8
adjust-cursor-height = 8
```

And, here are the contents of `theme/catppuccin-mocha`:
```
theme = catppuccin-mocha
```

And, here are the contents of `theme/tokyonight-moon`:
```
theme = /Users/kaz/repos/tokyonight.nvim/extras/ghostty/tokyonight_moon
```

With all the snippets defined, I then define the default for each grouping in my main Ghostty `config` file via the three `config-file` keys at the bottom:
```
font-size = 14
macos-option-as-alt = true
mouse-hide-while-typing = true
config-file = font/berkeley-mono
config-file = theme/tokyonight-moon
config-file = opacity/100%
```

Finally, here is the `gc` function defined in my `.zshrc`, which uses `fd` to feed the list of snippets to `fzf` excluding the main Ghostty `config` file. Then I bind an action to `fzf` and use the delimiter option to break apart the `dir/snippet` into fields `{1}` and `{2}` that I use in a `sed` statement to replace the correct `config-file` key in my main config. Lastly, because I'm on a Mac, I can use the native integration of Ghostty to call AppleScript to reload my configuration.
```bash
function gc {
  local GHOSTTY_DIR="$HOME/.config/ghostty"
  local CMD="sed -i '' 's:\(config-file = {1}\)/.*:\1/{2}:' $GHOSTTY_DIR/config && osascript -so -e 'tell application \"Ghostty\" to activate' -e 'tell application \"System Events\" to keystroke \",\" using {command down, shift down}'"
  fd \
    --type f \
    --exclude 'config' \
    --base-directory $GHOSTTY_DIR \
  | fzf \
    --preview "cat $GHOSTTY_DIR/{}" \
    --delimiter=/ \
    --bind="enter:become:$CMD"
}
```

I hope others find this useful.

**Comment by pkazmier** [comment]
The next step was to bind this to a key that would pop open the quick term and then feed in the `gc` command to run, but I don't think it's possible as you cannot compose actions it seems.

**Comment by zhenlonghe** [comment]
@pkazmier this is really awesome, thx for sharing.

**Reply by pkazmier** [reply]
You're welcome! I couldn't live without it as I switch fonts **constantly** 😄

**Reply by zhenlonghe** [reply]
Aha, we have the same interests in fonts! By the way, your nvim configuration is also very elegant!

**Reply by zhenlonghe** [reply]
Hey, Bro! I built on your work to add command arguments and autocompletion. =)

```shell
# ghostty config
function ghc {
  local GHOSTTY_DIR="$HOME/.config/ghostty"
  local CMD="sed -i '' 's:\(config-file = {1}\)/.*:\1/{2}:' $GHOSTTY_DIR/config && osascript -so -e 'tell application \"Ghostty\" to activate' -e 'tell application \"System Events\" to keystroke \",\" using {command down, shift down}'"

  local selected
  if [ -n "$1" ]; then
    selected="$1"
    if [ ! -f "$GHOSTTY_DIR/$selected" ]; then
      echo "File not found: $selected"
      return 1
    fi
  else
    selected=$(fd \
      --type f \
      --exclude 'config' \
      --base-directory $GHOSTTY_DIR \
    | fzf \
      --reverse \
      --preview "cat $GHOSTTY_DIR/{}" \
      --delimiter=/ \
      --bind="enter:become(echo {})")
    if [ -z "$selected" ]; then
      return 0
    fi
  fi

  local token1 token2 rest
  IFS=/ read -r token1 token2 rest <<< "$selected"
  if [ -n "$rest" ]; then
    echo "Path depth exceeds limit; only two levels supported (e.g., subdir/file)"
    return 1
  fi
  if [ -z "$token2" ]; then
    token2="$token1"
    token1=""
  fi

  local cmd=${CMD//\{1\}/$token1}
  cmd=${cmd//\{2\}/$token2}
  eval "$cmd"
}

_ghc_complete() {
  local GHOSTTY_DIR="$HOME/.config/ghostty"
  local files=($(fd --type f --exclude 'config' --base-directory "$GHOSTTY_DIR" 2>/dev/null))
  _describe 'files' files
}
compdef _ghc_complete ghc
```

**Reply by pkazmier** [reply]
Nice!

**Comment by blaugi** [comment]
Hey there, pretty cool setup. I'm trying to setup commit mono as my font but it doesnt seem to recognize it even though its installed. Can i see whats in your commit mono config?

**Reply by pkazmier** [reply]
Sure: https://github.com/pkazmier/ghostty-config/blob/main/font/commit-mono

**Comment by codeRSH** [comment]
Thanks copied your config files :)

**Reply by pkazmier** [reply]
Glad they helped!

**Comment by davidsanchez222** [comment]
+1 this is really cool, i'm going to add this to my dotfiles

**Reply by pkazmier** [reply]
Thanks. I couldn’t live without it coming from wezterm where I had built something similar (actually better because wezterm uses Lua for configuration).

**Comment by mitchellh** [comment, marked as answer]
Thanks for sharing. :smile:
